pkg/cloudprovider/providers/openstack/openstack.go

```diff
@@ -520,12 +521,13 @@ func getAddressByName(client *gophercloud.ServiceClient, name types.NodeName) (s
 	}
 
 	for _, addr := range addrs {
-		if addr.Type == v1.NodeInternalIP {
+		isIPv6 := net.ParseIP(addr.Address).To4() == nil
+		if (addr.Type == v1.NodeInternalIP) && (isIPv6 == needIPv6) {
 			return addr.Address, nil
 		}
 	}
 
-	return addrs[0].Address, nil
+	return "", ErrNoAddressFound
```

Comment:
With this last change, this function will _only_ return InternalIPs.  It should instead try ExternalIPs before giving up.

Aside: An alternate version might be to pass an address family down to `getAddressesByName` and thence `nodeAddresses` (with some value to indicate "all" families) and then do the filtering at the lowest level, when we know the address family from the openstack data structure.  Happy to go with whatever version you think is best.